twilight-gateway/src/inflater.rs

```diff
@@ -121,197 +121,97 @@ impl Inflater {
 
         // Decompressed message. `Vec::extend_from_slice` efficiently allocates
         // only what's necessary.
         let mut decompressed = Vec::new();
 
         loop {
-            let produced_pre = self.produced();
+            let mut output = OutBuffer::around(self.buffer.as_mut());
 
-            // Use Sync to ensure data is flushed to the buffer.
-            self.decompress
-                .decompress(
-                    &message[processed..],
-                    &mut self.buffer,
-                    FlushDecompress::Sync,
-                )
-                .map_err(|source| CompressionError {
-                    kind: CompressionErrorType::Decompressing,
-                    source: Some(Box::new(source)),
-                })?;
+            self.ctx
+                .decompress_stream(&mut output, &mut input)
+                .map_err(CompressionError::from_code)?;
 
-            processed = (self.processed() - processed_pre).try_into().unwrap();
-            let produced = (self.produced() - produced_pre).try_into().unwrap();
-
-            decompressed.extend_from_slice(&self.buffer[..produced]);
+            decompressed.extend_from_slice(output.as_slice());
 
             // Break when message has been fully decompressed.
-            if processed == message.len() {
+            if input.pos == input.src.len() && output.pos() != output.capacity() {
                 break;
             }
-
-            tracing::trace!(bytes.compressed.remaining = message.len() - processed);
-        }
-
-        {
-            #[allow(clippy::cast_precision_loss)]
-            let total_percentage_compressed =
-                self.processed() as f64 * 100.0 / self.produced() as f64;
-            let total_percentage_saved = 100.0 - total_percentage_compressed;
-            let total_kib_saved = (self.produced() - self.processed()) / 1024;
-
-            tracing::trace!(
-                bytes.compressed = message.len(),
-                bytes.decompressed = decompressed.len(),
-                total_percentage_saved,
-                "{total_kib_saved} KiB saved in total",
-            );
         }
 
-        self.clear();
+        self.processed += u64::try_from(input.src.len()).unwrap();
+        self.produced += u64::try_from(decompressed.len()).unwrap();
 
-        String::from_utf8(decompressed)
-            .map(Some)
-            .map_err(|source| CompressionError {
-                kind: CompressionErrorType::NotUtf8,
-                source: Some(Box::new(source)),
-            })
+        String::from_utf8(decompressed).map_err(|source| CompressionError {
+            kind: CompressionErrorType::NotUtf8,
+            source: Some(Box::new(source)),
+        })
     }
 
     /// Reset the inflater's state.
     pub(crate) fn reset(&mut self) {
-        self.compressed = Vec::new();
-        self.decompress.reset(true);
+        self.ctx
+            .reset(ResetDirective::SessionOnly)
```

Comment:
I think this should be `SessionAndParameters`, Discord does not reuse dictionaries between sessions as far as I know.

Reply:
Seemingly it is not possible to reset the internal hashset of zstd, so we should assume that it is fine to just reset the session.